File: Week 2/Week 2_ Solutions/C/C_freeweights.java
import java.io.*;
import java.util.*;

// make sure you submit w/ the class name set to Main if you use this template
class Main{
    static boolean validate(int n, ArrayList<Integer> a, ArrayList<Integer> b, int heaviestWeight){
        // remove all weights we can; we can always pair these together
        // now we need to check if the weights we can't move are paired
        ArrayList<Integer> remainingA = new ArrayList<Integer>();
        ArrayList<Integer> remainingB = new ArrayList<Integer>();
        for(int i : a){
            if(i > heaviestWeight){
                remainingA.add(i);
            }
        }
        for(int i : b){
            if(i > heaviestWeight){
                remainingB.add(i);
            }
        }
        if(remainingA.size() % 2 == 1) return false; // can't pair an odd number of weights
        if(remainingB.size() % 2 == 1) return false;
        for(int i = 0; i < remainingA.size(); i += 2){ 
            // don't use != on Objects
            if(!remainingA.get(i).equals(remainingA.get(i + 1))) return false; // these elements are unpaired
        }
        for(int i = 0; i < remainingB.size(); i += 2){
            if(!remainingB.get(i).equals(remainingB.get(i + 1))) return false;
        }
        return true;
    }

    static int freeWeights(int n, ArrayList<Integer> a, ArrayList<Integer> b){
        int lo = 0, hi = 1000000015; // 'hi' must be >= the max weight
        while(lo < hi){
            int mid = (lo + hi) / 2;
            if(validate(n, a, b, mid)){
                hi = mid;
            }else{
                lo = mid + 1;
            }
        }
        return lo;
    }
    public static void main(String[] args) throws IOException{
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(in.readLine());
        ArrayList<Integer> a = new ArrayList<Integer>();
        ArrayList<Integer> b = new ArrayList<Integer>();
        
        StringTokenizer st = new StringTokenizer(in.readLine());
        for(int i = 0; i < n; i++){
            a.add(Integer.parseInt(st.nextToken()));
        }
        st = new StringTokenizer(in.readLine());
        for(int i = 0; i < n; i++){
            b.add(Integer.parseInt(st.nextToken()));
        }
        
        System.out.println(freeWeights(n, a, b));
    }
}

File: Week 2/Week 2_ Solutions/C/C_freeweights.cpp
#include <bits/stdc++.h> 
using namespace std;
                                                // heaviest weight we can carry
bool validate(int n, vector<int>& a, vector<int>& b, int heaviest_weight){
    // remove all weights we can; we can always pair these together
    // now we need to check if the weights we can't move are paired
    vector<int> remaining_a, remaining_b;
    for(int i = 0; i < n; i++){
        if(a[i] > heaviest_weight){
            remaining_a.push_back(a[i]);
        }
        if(b[i] > heaviest_weight){
            remaining_b.push_back(b[i]);
        }
    }
    if(remaining_a.size() % 2 == 1) return false; // can't pair an odd number of weights
    if(remaining_b.size() % 2 == 1) return false;
    for(int i = 0; i < remaining_a.size(); i += 2){
        if(remaining_a[i] != remaining_a[i + 1]) return false; // these weights aren't paired
    }
    for(int i = 0; i < remaining_b.size(); i += 2){
        if(remaining_b[i] != remaining_b[i + 1]) return false;
    }
    return true;
}
int free_weights(int N, vector<int>& a, vector<int>& b){
    int lo = 0, hi = 1e9 + 5; // 'hi' must be greater/equal than the maximum weight
    while(lo < hi){
        int mid = (lo + hi) / 2;
        if(validate(N, a, b, mid)){
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    return lo;  
}

int main(){
    int n;
    cin >> n;
    vector<int> a(n), b(n);
    for(int i = 0; i < n; i++){
        cin >> a[i];
    }
    for(int i = 0; i < n; i++){
        cin >> b[i];
    }
    cout << free_weights(n, a, b) << endl;
}   